torch/testing/_internal/distributed/rpc/dist_autograd_test.py

```diff
@@ -2101,6 +2101,43 @@ def post_hook_add_two(output_grads, input_grads):
             self.assertEqual(1, len(grads))
             self.assertTrue(t in grads)
 
+    @staticmethod
+    def _slow_add(t1, t2):
+        time.sleep(1)
+        t3 = (t1 + t2).detach()
+        t3.requires_grad = True
+        return t3
+
+    @dist_init
+    def test_thread_local_context_id(self):
+        t1 = torch.rand((3, 3), requires_grad=True)
+        t2 = torch.rand((3, 3), requires_grad=True)
+
+        t3 = (t1 + t2).detach()
+        t3.requires_grad = True
+        t3.sum().backward()
+
+        dst = worker_name((self.rank + 1) % self.world_size)
+        rref = rpc.remote(dst, DistAutogradTest._slow_add, args=(t1, t2))
+
+        with dist_autograd.context() as context_id:
+            loss = rref.to_here().sum()
```

Comment:
Let me write down some notes for someone who is like me not super clear with FORWARD_AUTOGRAD_REQ.

The `rref.to_here()` here is a `PYTHON_RREF_FETCH_CALL`, which is always wrapped in `FORWARD_AUTORAD_REQUEST`, because of `forceGradRecording` on this line.

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/rref_impl.cpp#L145-L149

When the `PYTHON_RREF_FETCH_RESPONSE` is ready, as in this line.

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/request_callback_impl.cpp#L401

The response message be wrapped into `FORWARD_AUTOGRAD_RESP` by this line. 

https://github.com/pytorch/pytorch/blob/dfcea820df987cb12a68659c15747ee32e31ddad/torch/csrc/distributed/rpc/request_callback_impl.cpp#L496-L499

Where the critical `getMessageWithAutograd(...)` is called to add a SendFunction to the autograd context and the on the client receiving this `FORWARD_AUTOGRAD_RESP ` will add a corresponding RecvFunction to the dist autograd context.

If the autograd context is not restored, the SendFunction will be added to a wrong autograd context. and the following `sum()` backward will be propagated back to server but in a wrong autograd context. 

Why this test can reproduce the thread switch?

By making `rpc.remote(..., _slow_add, ...)` request slow, and the `rref.to_here()` here is processed before the RRef value is set.

So the `to_here()` request callback will run on another thread. The thread to run it is exactly the thread for `rpc.remote(..)`, because it marks the value as ready, thus responsible to run the added callbacks. 
Since `rpc.remote(..)` is called outside of a dist autograd context, `getMessageWithAutograd(...)` does not wrap the `PYTHON_REMOTE_CALL` into a `FORWARD_AUTORAD_REQUEST`, no autograd context is available on this thread.